hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java

```diff
@@ -77,6 +79,20 @@ public Boolean eval(Row row, ExpressionEvalContext context) {
             return null;
         }
 
+        if (this.operand1.getType().getTypeFamily() == QueryDataTypeFamily.OBJECT) {
+            Class<?> leftClass = left.getClass();
+            Class<?> rightClass = right.getClass();
+
+            if (!leftClass.equals(rightClass)) {
+                throw QueryException.error("Cannot compare two OBJECT values, because they have different classes");
```

Comment:
consider printing the actual class. 
more diagnostics could be very useful here. for example with Portable object it would be good to know its class/factory IDs. 

otherwise it's hard to tell what actually went wrong and how to fix it.